File: CusipPriceProcessor/CusipProcessor.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Utils;

namespace CusipPriceProcessor
{
	public class CusipProcessor
	{
		private readonly StreamReader _feedStream;
		private int _currentLineCount = 0;

		public CusipProcessor(StreamReader feedStream)
		{
			_feedStream = feedStream;
		}

		/// <summary>
		/// This method will be used for the first line of the feed file to be read which is expected to be a cusip
		/// </summary>
		/// <returns></returns>
		public string ReadCusip()
		{
			if (_feedStream.EndOfStream) throw new FeedException("Feed file seems to be empty", 1, string.Empty);

			// Reading asyncronously of a single line may be an overkill, it will be useful if the file is in a remote location
			// In which case we will need to provide a timeout mechanism.  Since decorating the method as async and returning a Task<string>
			// I decided that this is a half baked solution and as such we will read the file synchronously as opposed to the decoration and
			// The read as follows:
			//		var line = await _feedStream.ReadLineAsync().ConfigureAwait(false);
			var line = _feedStream.ReadLine();
			++_currentLineCount;
			var cusip = line.GetCusip();
			if (cusip == null) throw new FeedException("First line of Feed is not a cusip", _currentLineCount, line);

			return cusip;
		}

		/// <summary>
		/// At the beginning of this routine we are at the state where a Cusip was read
		/// Therefore, the next state is either a Price, EOF or an Exception.
		/// 
		/// This is a Left-to-Right parsing of the feed file, going through the file a line at a
		/// time with a single line lookahead, LL(1).
		/// </summary>
		/// <param name="cusip"></param>
		/// <returns></returns>
		public IEnumerable<CusipLatestPrice> ReadPricesForCusips(string cusip)
		{
			// This isPreviousBlank can be considered an additional state in the state machine I discussed in the
			// "Assumptions and discussion.txt" file
			var isPreviousBlank = false;

			// Loop through the cusips
			for (; ; )
			{
				if (_feedStream.EndOfStream) yield break;

				// A place holder for the fact that we encountered no price yet
				var prevPrice = decimal.MinValue;

				// Loop through prices
				for (; ; )
				{
					if (_feedStream.EndOfStream)
					{
						if (prevPrice == decimal.MinValue) throw new FeedException($"Cusip {cusip} contains no price value", _currentLineCount);
						yield return new CusipLatestPrice(cusip, prevPrice);
						break;
					}

					++_currentLineCount;

					// In the event that this statement throws an AggregateException then it will be caught by one of the calling routines
					var line = Task.Run(_feedStream.ReadLineAsync).Result;

					// It is acceptable for the last lines to be blank lines
					if (string.IsNullOrEmpty(line))
					{
						isPreviousBlank = true;
						continue;
					}
					else if (isPreviousBlank)
					{
						throw new FeedException("A blank line in the middle of the file indicates a corrupt Feed file", _currentLineCount);
					}

					var price = line.GetPrice();
					if (!price.HasValue)
					{
						// At this point we finished reading the prices for a cusip
						// If the cusip contained no prices then the feed file is corrupt
						// Since the line is not a price then it is either a Cusip or the feed file is corrupt
						var nextPrice = line.GetCusip();
						if (nextPrice == null) throw new FeedException($"Line is neither a cusip nor a price", _currentLineCount, line);
						if (prevPrice == decimal.MinValue) throw new FeedException($"Cusip {cusip} contains no price value", _currentLineCount, line);

						yield return new CusipLatestPrice(cusip, prevPrice);

						cusip = nextPrice;
						break;
					}

					prevPrice = price.Value;
				}
			}
		}

		/// <summary>
		/// Since we initalize the EventHandler there is no need to check for null when we
		/// raise the event.
		/// </summary>
		public event EventHandler<CusipPriceEventArgs> CusipPriceHandler = delegate { };

		public async Task<bool> ProcessFeedFile()
		{
			var t = await Task.Run(() => {
				var cusip = ReadCusip();
				var cusipPrices = ReadPricesForCusips(cusip);
				foreach (var clp in cusipPrices)
					Raise(new CusipPriceEventArgs(clp.Cusip, clp.Price));

				return true;
			});

			return t;
		}

		private void Raise(CusipPriceEventArgs ea) => CusipPriceHandler(this, ea);
	}
}

File: CusipPriceProcessorCpp/CusipProcessor.cpp
#include "pch.h"
#include <limits>
#include <sstream>
#include <regex>
#include "CusipProcessor.h"
#include "..\\UtilsCpp\FeedExceptionCpp.h"
#include "..\\UtilsCpp\StringExt.h"

using namespace std;

static regex CusipReCusipProcessor("^\\s*(\\w{8})\\s*$");
static regex PriceReCusipProcessor("^\\s*(\\d*\\.\\d+)\\s*$");
static string emptyNextCusipCusipProcessor = string("");

regex& CusipProcessor::cusipRe(CusipReCusipProcessor);
regex& CusipProcessor::priceRe(PriceReCusipProcessor);

CusipProcessor::CusipProcessor(ifstream& feedStr) : feedStream(feedStr), currentLineCount(0), NextCusip(emptyNextCusipCusipProcessor), PreviousPrice(0.0)
{
}

CusipProcessor::~CusipProcessor()
{
}

string CusipProcessor::ReadCusip()
{
	if (feedStream.eof()) return string("");

	string cusip;
	getline(feedStream, cusip);
	++currentLineCount;
	if (cusip.empty())
		throw FeedExceptionCpp(string("First line of Feed is not a cusip"), currentLineCount, cusip);

	NextCusip = cusip;
	return cusip;
}

double CusipProcessor::GetPrice(string line)
{
	smatch matches;
	regex_search(line, matches, priceRe);
	if (matches.empty())
		return numeric_limits<double>::lowest();

	auto sPrice = matches.str(1);
	auto price = atof(sPrice.c_str());
	return price;
}

string CusipProcessor::GetCusip(string line)
{
	smatch matches;
	regex_search(line, matches, cusipRe);
	if (matches.empty())
		return emptyNextCusipCusipProcessor;

	auto cusip = matches.str(1);
	return cusip;
}


CusipLatestPrice CusipProcessor::ReadPricesForCusips()
{
	auto isPreviousBlank = false;
	string cusip = NextCusip;

	// Loop through the cusips
	for (; ; )
	{
		if (feedStream.eof()) return CusipLatestPrice::NoValue;

		// Loop through prices
		for (; ; )
		{
			if (feedStream.eof())
			{
				if (PreviousPrice == numeric_limits<double>::lowest())
				{
					stringstream msg;
					msg << "Cusip " << cusip << " contains no price value";
					throw FeedExceptionCpp(msg.str(), currentLineCount);
				}
				return CusipLatestPrice(cusip, PreviousPrice);
			}

			string line = string("");
			if (feedStream.peek() == '\n')
			{
				istreambuf_iterator<char> iter(feedStream);
				++iter;
				++currentLineCount;
			}
			else if (feedStream.peek() == '\r')
			{
				// Advance the file by 1 char and do not inc line count
				istreambuf_iterator<char> iter(feedStream);
				++iter;
				continue;
			}
			else
			{
				getline(feedStream, line);
				++currentLineCount;
			}

			// It is acceptable for the last lines to be blank lines
			if (Trim(line) == string(""))
			{
				isPreviousBlank = true;
				continue;
			}
			else if (isPreviousBlank)
			{
				throw FeedExceptionCpp("A blank line in the middle of the file indicates a corrupt Feed file", currentLineCount, line);
			}

			auto price = GetPrice(line);
			if (price == numeric_limits<double>::lowest())
			{
				auto nextCusip = GetCusip(line);
				if (nextCusip == string(""))
				{
					throw FeedExceptionCpp("Line is neither a cusip nor a price", currentLineCount, line);
				}
				if (PreviousPrice == numeric_limits<double>::lowest())
				{
					stringstream msg;
					msg << "Cusip " << cusip << " contains no price value";
					throw FeedExceptionCpp(msg.str(), currentLineCount, line);
				}

				auto prevCusip = cusip;
				NextCusip = nextCusip;
				price = PreviousPrice;
				PreviousPrice = numeric_limits<double>::lowest();
				return CusipLatestPrice(prevCusip, price);
			}

			PreviousPrice = price;
		}
	}
}

